### PR TITLE
add deterministic default cost

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

Default instruction cost to `ComputeBudget::get_max_units()` is safer than defaulting to `0`.

#### Summary of Changes

- Add `get_default_units()`, use it instead of `0` as default cost units for unknown instructions

Fixes #

### CODE REVIEW

**Review by jackcmay — APPROVED** [review]

**Comment by taozhu-chicago** [comment]
The 3 failed tests is waiting other PRs. I'll update this when merged #23645

**Comment by carllin** [comment]
hmmm I don't see this function being used outside of tests. Can you provide some context/intended use cases where this is a safer "average" than `0` when the table is empty?

**Comment by t-nelson** [comment]
> hmmm I don't see this function being used outside of tests. Can you provide some context/intended use cases where this is a safer "average" than `0` when the table is empty?

I'd also like to hear a justification for this value.  Using max seems _very_ confident that we don't have an off-by-one error :sweat_smile: 

`max / 2` seems a little more in line

**Comment by taozhu-chicago** [comment]
> > hmmm I don't see this function being used outside of tests. Can you provide some context/intended use cases where this is a safer "average" than `0` when the table is empty?
> 
> I'd also like to hear a justification for this value. Using max seems _very_ confident that we don't have an off-by-one error 😅
> 
> `max / 2` seems a little more in line

The use case is when it needs to provide cost for a new instruction, at [here](https://github.com/solana-labs/solana/pull/23656/files#diff-05b5f07ae482624062494888162a725444b51ee716f1f147affef29b785386c7R136-R141). For a unknown instruction, i went through few options: 
- using average units amount all instructions, `get_average_units()`, which I found being insufficient as instructions cost units vary greatly. `Average` doesn't really have any underlying theory to support it. 
- using the cost of most frequently used instruction, `get_mode_units()`, the idea is the most "popular" instruction in cluster could be a good reference. The downside is most frequently used instruction are usually the cheap ones. Plus, over-packing a block became the major issue. 
- That leads to use `get_default_units()` that:
1. better have a deterministic default units for unknown instruction (so out the average and mode)
2. safer to assume unknown instruction as expensive (hence 200_000) until prove otherwise.

As for the actual value for `max_unit`, I'm under impression that `1_400_000` is a safe limit for white-listed instructions, I might be wrong tho. 

However the concern of starting with `1_400_000` is, if the instruction actually takes much lower units, as normally they would be, then it'd take few more data point for EMA (or any average algo) to stabilize. 

`max / 2` would work too, but still think the `/2` is arbitrary.

File: runtime/src/execute_cost_table.rs [hunk 1]
```diff
@@ -45,19 +47,25 @@ impl ExecuteCostTable {
         self.table.len()
     }
 
-    // instead of assigning unknown program with a configured/hard-coded cost
-    // use average or mode function to make a educated guess.
-    pub fn get_average(&self) -> u64 {
+    // default program cost to max_units
```
Comment (carllin) [inline runtime/src/execute_cost_table.rs:50]: `///` for code comments

Also this comment doesn't seem that useful 😃

File: runtime/src/execute_cost_table.rs [hunk 2]
```diff
@@ -45,19 +47,25 @@ impl ExecuteCostTable {
...
+    pub fn get_default_units(&self) -> u64 {
+        let use_max_units_default = false;
+        ComputeBudget::get_max_units(use_max_units_default)
+    }
+
+    // average cost of all recorded programs
```
Comment (carllin) [inline runtime/src/execute_cost_table.rs:56]: nit: `///` for function documentation

File: core/src/cost_update_service.rs [hunk 3]
```diff
@@ -264,20 +248,47 @@ mod tests {
...
+                    .get(&program_key_1)
+            );
         }
 
         // Test updating cost model with only erroring compute costs where the `cost_per_error` is
         // greater than the current instruction cost for the program. Should update with the
         // new erroring compute costs
         let cost_per_error = 1000;
+        // the expect cost is (previous_cost + new_cost)/2 = (100 + 1000)/2 = 550
+        let expect_units = 550;
```
Comment (carllin) [inline core/src/cost_update_service.rs:291]: nit: `expect_units -> expected_units`

File: program-runtime/src/compute_budget.rs [hunk 4]
```diff
@@ -102,6 +97,14 @@ impl ComputeBudget {
         }
     }
 
+    pub fn get_max_units(use_max_units_default: bool) -> u64 {
+        if use_max_units_default {
+            MAX_UNITS as u64
+        } else {
+            200_000
```
Comment (carllin) [inline program-runtime/src/compute_budget.rs:104]: Can we make this a constant `DEFAULT_UNITS`

File: program-runtime/src/compute_budget.rs [hunk 5]
```diff
@@ -73,13 +73,8 @@ impl Default for ComputeBudget {
 
 impl ComputeBudget {
     pub fn new(use_max_units_default: bool) -> Self {
-        let max_units = if use_max_units_default {
-            MAX_UNITS
-        } else {
-            200_000
-        } as u64;
         ComputeBudget {
-            max_units,
```
Comment (carllin) [inline program-runtime/src/compute_budget.rs:82]: Can we change this such that the `ComputeBudget` just passes in the `max_units` instead of `use_max_units_default`. 

Then the constructor of `ComputeBudget` can either use one of:
1. `ComputeBudget::new(DEFAULT_UNITS)` (constant suggested by this comment here: https://github.com/solana-labs/solana/pull/23656/files#r827652115)
2. `ComputeBudget::new(MAX_UNITS)`

File: runtime/src/execute_cost_table.rs [hunk 6]
```diff
@@ -45,19 +47,25 @@ impl ExecuteCostTable {
         self.table.len()
     }
 
-    // instead of assigning unknown program with a configured/hard-coded cost
-    // use average or mode function to make a educated guess.
-    pub fn get_average(&self) -> u64 {
+    // default program cost to max_units
+    pub fn get_default_units(&self) -> u64 {
+        let use_max_units_default = false;
+        ComputeBudget::get_max_units(use_max_units_default)
```
Comment (carllin) [inline runtime/src/execute_cost_table.rs:53]: This is unintuitive to read because `get_max_units()` and `get_default_units()` seem unrelated, yet one calls the other.

I think if we follow this suggestion here: https://github.com/solana-labs/solana/pull/23656/files#r827652688, we can just directly use the constants instead of making function calls to indirectly get the constants

Comment (taozhu-chicago) [reply runtime/src/execute_cost_table.rs:53]: +1

**Comment by taozhu-chicago** [comment]
With the latest change to directly use compute_budget's MAX_UNITS and DEFAULT_UNITS, I feel the logic is cleaner, it becomes self-evident to use DEFAULT_UNITS as _default_ value for unknown instructions.  😅
@t-nelson @carllin can you give it another look? Thanks!

File: program-runtime/src/invoke_context.rs [hunk 7]
```diff
@@ -1659,20 +1660,20 @@ mod tests {
         let mut transaction_context = TransactionContext::new(accounts, 1, 3);
         let mut invoke_context = InvokeContext::new_mock(&mut transaction_context, &[]);
         invoke_context.feature_set = Arc::new(feature_set);
-        invoke_context.compute_budget = ComputeBudget::new(false);
+        invoke_context.compute_budget = ComputeBudget::new(DEFAULT_UNITS);
```
Comment (t-nelson) [inline program-runtime/src/invoke_context.rs:1663]: same self-doc suggestion in this file

File: runtime/src/bank.rs [hunk 8]
```diff
@@ -4036,9 +4036,14 @@ impl Bank {
                     signature_count += u64::from(tx.message().header().num_required_signatures);
 
                     let tx_wide_compute_cap = feature_set.is_active(&tx_wide_compute_cap::id());
+                    let compute_budget_max_units = if tx_wide_compute_cap {
+                        MAX_UNITS
```
Comment (t-nelson) [inline runtime/src/bank.rs:4040]: and here :slightly_smiling_face:

File: runtime/src/cost_model.rs [hunk 9]
```diff
@@ -113,9 +113,9 @@ impl CostModel {
         match self.instruction_execution_cost_table.get_cost(program_key) {
             Some(cost) => *cost,
             None => {
-                let default_value = self.instruction_execution_cost_table.get_mode();
+                let default_value = self.instruction_execution_cost_table.get_default_units();
                 debug!(
-                    "Program key {:?} does not have assigned cost, using mode {}",
+                    "Instruction {:?} does not have aggregated cost, using default value {}",
```
Comment (t-nelson) [inline runtime/src/cost_model.rs:118]: I think "program" was more accurate here?  The following format field is a program id and we have no way of reliably discriminating instructions in a general way

File: runtime/src/execute_cost_table.rs [hunk 10]
```diff
@@ -45,19 +47,24 @@ impl ExecuteCostTable {
         self.table.len()
     }
 
-    // instead of assigning unknown program with a configured/hard-coded cost
-    // use average or mode function to make a educated guess.
-    pub fn get_average(&self) -> u64 {
+    /// default program cost, set to ComputeBudget::DEFAULT_UNITS
+    pub fn get_default_units(&self) -> u64 {
```
Comment (t-nelson) [inline runtime/src/execute_cost_table.rs:51]: Is this getter necessary now?  Seems like we can just use the constant directly

File: runtime/src/execute_cost_table.rs [hunk 11]
```diff
@@ -45,19 +47,24 @@ impl ExecuteCostTable {
...
-            0
+            self.get_default_units()
         } else {
             self.table.iter().map(|(_, value)| value).sum::<u64>() / self.get_count() as u64
         }
     }
 
-    pub fn get_mode(&self) -> u64 {
+    /// the most frequently occurring program's cost
+    pub fn get_mode_units(&self) -> u64 {
```
Comment (t-nelson) [inline runtime/src/execute_cost_table.rs:65]: wdyt about changing these to something like `get_statistical_mode_units`, etc?  For whatever reason I didn't immediately jump to "statistical," but rather, "operating" mode.  I'd have probably misread or been confused by code at the call site

**Review by t-nelson — APPROVED** [review]
ok you've got me down to nits :slightly_smiling_face: 

thanks for the cleanup! it's much more clear now.

r+ nits

File: cli/src/cluster_query.rs [hunk 12]
```diff
@@ -1408,7 +1408,7 @@ pub fn process_ping(
             )];
             if let Some(additional_fee) = additional_fee {
                 ixs.push(ComputeBudgetInstruction::request_units(
-                    ComputeBudget::new(false).max_units as u32,
+                    DEFAULT_UNITS,
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply cli/src/cluster_query.rs]: Good call, ✅

File: program-runtime/src/invoke_context.rs (hunk 7 above)
Comment (taozhu-chicago) [reply program-runtime/src/invoke_context.rs]: ✅

File: runtime/src/bank.rs (hunk 8 above)
Comment (taozhu-chicago) [reply runtime/src/bank.rs]: ✅

File: runtime/src/cost_model.rs (hunk 9 above)
Comment (taozhu-chicago) [reply runtime/src/cost_model.rs]: Yea, been back-n-forth on this few times, `program` sometimes means a static code that does not associate with execution cost, until `instruction` to provide execution path, then comes `units` it'd consume. 

But pairing "Program" with program ID makes more sense.

**Comment by taozhu-chicago** [comment]
Cleaned up the remaining few bits, CI is happy too. @t-nelson, good to go?

**Review by carllin — APPROVED** [review]

**Review by t-nelson — APPROVED** [review]
Thanks for entertaining the nits! :shipit:
